### PR TITLE
[MBL-1755] Pledge Summary Table Total Amount Text

### PR DESCRIPTION
**Pull request by scottkicks**
# 📲 What

We also should be showing “Pledge without a reward” when a user pledges without a reward, instead of “Bonus support” and the total “Pledge amount”

# 🛠 How

This section of the pledge summary is managed by `NoShippingPledgeRewardsSummaryTotalViewController` for crowdfunding checkout and `PostCampaignPledgeRewardsSummaryTotalViewController` for late pledge checkout.

We need these classes to know if the backer is pledging without a reward so they can display either  “Pledge without a reward” or “Pledge amount” so, I'm passing the `baseReward`, which has a handy `isNoReward` helper, from the view model through to these classes.

The designs also show that, when pledging without a reward, the bonus support and reward table view rows should be hidden.

# 👀 See
### [Designs for reference](https://www.figma.com/design/z4faZM4BQwbyiDe7y8tbpP/Pledge-Redemption?node-id=47-4736&t=WYnhXc3kE1QWwkcn-4)
Simulator Screen Recording - iPhone 15 Pro 17 5 - 2024-09-25 at 13 31 36

# ✅ Acceptance criteria

- [x] When pledging without a reward, the checkout screen's pledge summary table only shows the total summary row with the label text, "Pledge without a reward".
- [x] Pledges with any other reward type are shown in the table as normal

### CODE REVIEW

File: Library/ViewModels/PledgeSummaryViewModel.swift [hunk 1]
```diff
@@ -76,6 +77,13 @@ public class PledgeSummaryViewModel: PledgeSummaryViewModelType,
     }
     .skipNil()
 
+    self.titleLabelText = pledgeHasNoReward
+      .map { hasNoRewards in
+        featureNoShippingAtCheckout() && hasNoRewards != nil && hasNoRewards == true
+          ? Strings.Pledge_without_a_reward()
+          : Strings.Pledge_amount()
+      }
```
Comment (scottkicks) [inline Library/ViewModels/PledgeSummaryViewModel.swift]: cleaned this up in a later commit

**Review by ifosli — DISMISSED** [review]
Added a couple comments; I'm confused by why the "no reward" bool is separate from the rest of the config data and I'm missing a snapshot test for crowdfunding checkout. Would you please fix/explain?

Also, since this string is longer, would you mind double-checking what the UI looks like for large font sizes or maybe in german? UI doesn't need to be perfect but it should be usable.

File: Library/ViewModels/PostCampaignPledgeRewardsSummaryViewModel.swift
[the remark this answers is not in the document]
Comment (scottkicks) [reply Library/ViewModels/PostCampaignPledgeRewardsSummaryViewModel.swift:1]: I'm refactoring this table in my current ticket. This should make it easier for these classes to use one version of this table.

**Comment by nativeksr** [comment]
### SwiftFormat found issues:

| File | Rules |
| ---- | ----- |
| Kickstarter-iOS/Features/PledgeView/Controllers/NoShippingPledgeViewControllerTests.swift:259:1 | warning: (trailingSpace) Remove trailing space at end of a line. |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/ViewModels/ConfirmDetailsViewModel.swift [hunk 2]
```diff
@@ -260,10 +260,15 @@ public class ConfirmDetailsViewModel: ConfirmDetailsViewModelType, ConfirmDetail
...
       projectAndConfirmationLabelHidden,
-      pledgeTotal
+      pledgeTotal,
+      rewards
     )
-    .map(unpack)
-    .map { project, confirmationLabelHidden, total in (project, total, confirmationLabelHidden) }
+    .map { projectAndConfirmationLabelHidden, pledgeTotal, rewards in
+      let (project, confirmationLabelHidden) = projectAndConfirmationLabelHidden
+      let pledgeHasNoReward = rewards.count == 1 && rewards.first?.isNoReward == true
```
Comment (stevestreza-ksr) [inline Library/ViewModels/ConfirmDetailsViewModel.swift:268]: Nit: Could use `pledgeHasNoRewards(rewards:)` here

File: Library/ViewModels/NoShippingConfirmDetailsViewModel.swift [hunk 3]
```diff
@@ -203,10 +203,15 @@ public class NoShippingConfirmDetailsViewModel: NoShippingConfirmDetailsViewMode
...
       projectAndConfirmationLabelHidden,
-      pledgeTotal
+      pledgeTotal,
+      rewards
     )
-    .map(unpack)
-    .map { project, confirmationLabelHidden, total in (project, total, confirmationLabelHidden) }
+    .map { projectAndConfirmationLabelHidden, pledgeTotal, rewards in
+      let (project, confirmationLabelHidden) = projectAndConfirmationLabelHidden
+      let pledgeHasNoReward = rewards.count == 1 && rewards.first?.isNoReward == true
```
Comment (stevestreza-ksr) [inline Library/ViewModels/NoShippingConfirmDetailsViewModel.swift:211]: Same nit: Could use `pledgeHasNoRewards(rewards:)` here

File: Library/ViewModels/PostCampaignPledgeRewardsSummaryViewModel.swift [hunk 4]
```diff
@@ -211,7 +211,7 @@ private func items(
 
   // MARK: Bonus
 
-  if let bonus = bonusAmount, bonus > 0 {
+  if let bonus = bonusAmount, bonus > 0, rewardItems.isEmpty == false {
```
Comment (stevestreza-ksr) [inline Library/ViewModels/PostCampaignPledgeRewardsSummaryViewModel.swift:214]: Sanity checking this shouldn't be doing that same `isNoReward` check here

**Review by stevestreza-ksr — APPROVED** [review]
A few minor comments but overall good to go 👍

File: Library/ViewModels/PostCampaignPledgeRewardsSummaryViewModel.swift (hunk 4 above)
Comment (scottkicks) [reply Library/ViewModels/PostCampaignPledgeRewardsSummaryViewModel.swift:214]: either or would work here actually.

**Review by stevestreza-ksr — APPROVED** [review]

**Pull request merged**